### DISCUSSION TITLE
How are CHAP credentials details stored?

### DISCUSSION THREAD

**Discussion by daniel-moulton** [open-iscsi/open-iscsi#448] [Q&A] [answered]
Hi all,

I’ve been looking into CHAP authentication for iSCSI and was wondering if anyone knows about how the credentials are stored. I've had a look through the source but wasn't able to find a concrete answer to the below question.

The passwords (seem) to just be stored in plaintext, both in iscsid.conf and if using the --show argument when viewing a discovery record using iscisadm.

If this is the case, is there a way to prevent the password being plaintext / viewable? Perhaps a way to hash the passwords instead (although assuming that would require external functionality)

Any insight or pointers would be appreciated thanks.

**Comment by gonzoleeman** [comment, marked as answer]
> Hi all,
> 
> I’ve been looking into CHAP authentication for iSCSI and was wondering if anyone knows about how the credentials are stored. I've had a look through the source but wasn't able to find a concrete answer to the below question.
> 
> The passwords (seem) to just be stored in plaintext, both in iscsid.conf and if using the --show argument when viewing a discovery record using iscisadm.

Correct. The name and password  (both equally important) are stored in plaintext, but in a root-protected directory and file.

> 
> If this is the case, is there a way to prevent the password being plaintext / viewable? Perhaps a way to hash the passwords instead (although assuming that would require external functionality)
> 
> Any insight or pointers would be appreciated thanks.

You can not normally view the name and password using iscsiadm:

```
root@tumbleweed(1):~# iscsiadm -m node --op update --name node.session.auth.username --value fred
root@tumbleweed(1):~# iscsiadm -m node --op update --name node.session.auth.password -value secret01
root@tumbleweed(1):~# iscsiadm -m node --op show | egrep 'node.session.auth.(username|password) '
node.session.auth.username = fred
node.session.auth.password = ********
root@tumbleweed(1):~# 

```

As you can see, the code tries not to display the password. But of course anyone with root access can just look at the plaintext file. Plus, you can supply the "-S" flag and see the password, if you wish:
```
root@tumbleweed(1):~# iscsiadm -m node --op show -S | egrep 'node.session.auth.(username|password) '
node.session.auth.username = fred
node.session.auth.password = secret01
root@tumbleweed(1):~# 

```

By the way, even hashing the passwords adds limited security. But if it requires root to see the password, that means the attacker already has root access, and all security is lost.

**Reply by daniel-moulton** [reply]
Thanks for the quick response Lee, agree that if an attacker has root access to the system then being able to view the CHAP credentials is maybe not the biggest issue at large! Thanks for the help

**Comment by gonzoleeman** [comment]
Closing since this has been answered.
